### ISSUE TITLE
connection->enterEventLoop() doesn't block in a thread

### ISSUE THREAD

**Issue by myabhiasati** [Kistler-Group/sdbus-cpp#124]
Hi,
I am not sure if this is the right forum to ask this but posting here as I didn't find any group/forum for same. I am new to sdbus-c++ and completely blocked in my work due to below scenario. It would be of great help if you can help me resolve this or provide few pointers. I have included code snippet below.

**ISSUE**
I am trying to create an event loop but finding that it is not blocking for me i.e. I don't find my service registered in busctl when I create it in a thread like this (to be more specific, sometimes it appears momentorily and then disappears)
    std::thread t([&](){ connection->enterEventLoop(); });

If I use it directly like this, then it appears in busctl) but problem is it blocks the process from where I am calling this function all the later functions of the process are not executed.
   connection->enterEventLoop();

I tried Async loop also but it didn't help (no service appears in busctl):
  connection->enterEventLoopAsync();

Can you please tell if I am missing anything in starting the event loop properly?

**CODE SNIPPET**

void Create_Emitter_Connection(std::string emitter){

		const char * EserviceName = emitter.c_str();
		auto connection = sdbus::createSessionBusConnection(EserviceName);
		const char* objectPath = "/com/xyz/pqr/abc";
		auto emit = sdbus::createObject(*connection, objectPath);
		g_emit = emit.get();
		const char* interfaceName = "com.xyz.pqr.abc.Events1";
		
		emit->registerMethod("concatenate").onInterface(interfaceName).implementedAs(&concatenate);
		emit->finishRegistration();
		
                 //IT DOESN"T WORK IN EITHER OF THESE WAYS- I CAN SEE SERVICE NAME APPEARING MOMENTORILY IN BUSCTL AND THEN IT GOES AWAY
                 std::thread t([&](){ connection->enterEventLoop(); });
                 connection->enterEventLoopAsync();
		
                // IT WORKS LIKE THIS BUT PROBLEM IS IT BLOCKS THE PROCESS
		connection->enterEventLoop();
}

Thanks a lot.

**Comment by sangelovic** [comment]
Hello. What is a problem for you is actually the way it's supposed to work :-) So `enterEventLoop()` is supposed to block the thread. It's gonna listen to and process the D-Bus messages. It's a typical case for D-Bus services that they do the setup and then relinquish the control to D-Bus event loop. But, for your flexibility, if you don't want to block your main thread, you have an option to call `enterEventLoopAsync()` instead. Which creates a thread for you and runs event loop in that thread. **But** what you do wrong is that your main thread continues right away to the end of the `main()` function, causing the destruction of the `connection`, which in turn jumps out of the event loop, destroys the event loop thread, unregisters itself completely from D-Bus daemon, and your application finishes. If you want to use `enterEventLoopAsync()`, you have to have another way of making your main thread busy (perhaps handling some other requests in your application or polling on some event fd or whatever), otherwise your application will end right after it has started. Hence your very short observation of the service appearing in busctl.

So what you want is no `std::thread t([&](){ connection->enterEventLoop(); });`. And no `connection->enterEventLoopAsync();` where the main thread is going to finish right away. You want either `connection->enterEventLoop();` in your main thread, which will block it (and you can unblock it e.g. by SIGKILL signal handler which will call `connection->leaveEventLoop();`), or you want `connection->enterEventLoopAsync();`, but then your main thread must, logically, stay in `main()` somehow. Or you may want to get the sdbus fd and poll events and do the polling on your own.

**Comment by myabhiasati** [comment]
Hi,
Thanks a lot for your detailed answer. Now I have ensured that the main thread (my application) stay (by putting it in a while loop and waiting for signals. I have verified that my application is running by checking process list).

However, I still don't see service registered via busctl command. I have used this code:
    auto connection = sdbus::createSessionBusConnection(serviceName);
    connection->enterEventLoopAsync();

When I use enterEventLoop() I can see my service registered but I don't want to do that as it blocks at that point.

Can you please tell what else am I missing while creating Async loop?

Thanks

**Comment by myabhiasati** [comment]
Just to update this thread, I have found the issue.

Problem was my connection object was was going out of scope invoking it's destructor and hence deregistering the service. Declaring connection object at right place (as of now as a global variable) have unblocked me. Though I would not prefer making it global as a permanent solution.

**Comment by sangelovic** [comment]
Yes, that's what I also tried to explain you in my first reply. The
connection object must of course stay alive for the entire time. Global
instance is not the only option. You can have it as a member of a class
representing an application, for example. Or you keep it on the stack, but
then you gotta use nonasync enterEventLoop.

**Comment by vk-gst** [comment]
1. A question regarding the `connection->enterEventLoop();` and `connection->enterEventLoopAsync();`:
My server application is intended to receive multiple calls to my exported methods/interfaces, but however my server methods inturn call blocking methods(platform APIs). In such a case, I plan to use the asynchronous example(provided in docs) + synchronizing for blocking calls, is this the right approach? Or is sdbus-c++ capable of queuing up multiple requests issued by client, and can block so that I can attend the requests sequentially in the server?
what kind of eventloop I must be using in this case? It is a little confusing to understand the `connection->enterEventLoop();` and `connection->enterEventLoopAsync();` 


2. Additionally the asynchronous server example provided has following errors during compilation. Is this issue known already? Following are the logs:


```
g++ -std=c++1z  async-server.cpp -o aserver -lsdbus-c++ -pthread
In file included from /usr/local/include/sdbus-c++/Message.h:30:0,
                 from /usr/local/include/sdbus-c++/ConvenienceApiClasses.h:30,
                 from /usr/local/include/sdbus-c++/IObject.h:30,
                 from /usr/local/include/sdbus-c++/sdbus-c++.h:28,
                 from async-server.cpp:1:
/usr/local/include/sdbus-c++/TypeTraits.h: In instantiation of ‘static const string sdbus::signature_of<_T>::str() [with _T = sdbus::MethodCall; std::__cxx11::string = std::__cxx11::basic_string<char>]’:
/usr/local/include/sdbus-c++/TypeTraits.h:491:79:   required from ‘static const string sdbus::aggregate_signature<std::tuple<_Tps ...> >::str() [with _Types = {sdbus::MethodCall}; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/usr/local/include/sdbus-c++/TypeTraits.h:501:92:   required from ‘static const string sdbus::signature_of_function_input_arguments<_Function>::str() [with _Function = void (*)(sdbus::MethodCall); std::__cxx11::string = std::__cxx11::basic_string<char>]’
/usr/local/include/sdbus-c++/ConvenienceApiClasses.inl:93:80:   required from ‘sdbus::MethodRegistrator& sdbus::MethodRegistrator::implementedAs(_Function&&) [with _Function = void (*)(sdbus::MethodCall)]’
async-server.cpp:91:102:   required from here
/usr/local/include/sdbus-c++/TypeTraits.h:71:13: error: static assertion failed: Unknown DBus type
             static_assert(sizeof(_T) < 0, "Unknown DBus type");
             ^~~~~~~~~~~~~
In file included from /usr/local/include/sdbus-c++/ConvenienceApiClasses.h:30:0,
                 from /usr/local/include/sdbus-c++/IObject.h:30,
                 from /usr/local/include/sdbus-c++/sdbus-c++.h:28,
                 from async-server.cpp:1:
/usr/local/include/sdbus-c++/Message.h: In instantiation of ‘void sdbus::detail::deserialize_pack(sdbus::Message&, _Args& ...) [with _Args = {sdbus::MethodCall}]’:
/usr/local/include/sdbus-c++/Message.h:378:29:   required from ‘void sdbus::detail::deserialize_tuple(sdbus::Message&, _Tuple&, std::index_sequence<_Is ...>) [with _Tuple = std::tuple<sdbus::MethodCall>; long unsigned int ..._Is = {0}; std::index_sequence<_Is ...> = std::integer_sequence<long unsigned int, 0>]’
/usr/local/include/sdbus-c++/Message.h:402:34:   required from ‘sdbus::Message& sdbus::operator>>(sdbus::Message&, std::tuple<_Tps ...>&) [with _ValueTypes = {sdbus::MethodCall}]’
/usr/local/include/sdbus-c++/ConvenienceApiClasses.inl:102:18:   required from ‘sdbus::MethodRegistrator::implementedAs(_Function&&)::<lambda(sdbus::MethodCall)> [with _Function = void (*)(sdbus::MethodCall)]’
/usr/local/include/sdbus-c++/ConvenienceApiClasses.inl:95:71:   required from ‘struct sdbus::MethodRegistrator::implementedAs(_Function&&) [with _Function = void (*)(sdbus::MethodCall)]::<lambda(class sdbus::MethodCall)>’
/usr/local/include/sdbus-c++/ConvenienceApiClasses.inl:95:25:   required from ‘sdbus::MethodRegistrator& sdbus::MethodRegistrator::implementedAs(_Function&&) [with _Function = void (*)(sdbus::MethodCall)]’
async-server.cpp:91:102:   required from here
/usr/local/include/sdbus-c++/Message.h:370:13: error: no match for ‘operator>>’ (operand types are ‘sdbus::Message’ and ‘sdbus::MethodCall’)
             (void)(msg >> ... >> args);
/usr/local/include/sdbus-c++/Message.h:104:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(sdbus::Variant&) <near match>
         Message& operator>>(Variant &item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:104:18: note:   conversion of argument 1 would be ill-formed:
/usr/local/include/sdbus-c++/Message.h:370:13: error: cannot bind non-const lvalue reference of type ‘sdbus::Variant&’ to an rvalue of type ‘sdbus::Variant’
             (void)(msg >> ... >> args);
In file included from /usr/local/include/sdbus-c++/ConvenienceApiClasses.inl:34:0,
                 from /usr/local/include/sdbus-c++/IObject.h:494,
                 from /usr/local/include/sdbus-c++/sdbus-c++.h:28,
                 from async-server.cpp:1:
/usr/local/include/sdbus-c++/Types.h:59:9: note:   after user-defined conversion: sdbus::Variant::Variant(const _ValueType&) [with _ValueType = sdbus::MethodCall]
         Variant(const _ValueType& value)
         ^~~~~~~
In file included from /usr/local/include/sdbus-c++/ConvenienceApiClasses.h:30:0,
                 from /usr/local/include/sdbus-c++/IObject.h:30,
                 from /usr/local/include/sdbus-c++/sdbus-c++.h:28,
                 from async-server.cpp:1:
/usr/local/include/sdbus-c++/Message.h:370:13: note: candidate: operator>>(int, int) <built-in>
             (void)(msg >> ... >> args);
/usr/local/include/sdbus-c++/Message.h:370:13: note:   no known conversion for argument 2 from ‘sdbus::MethodCall’ to ‘int’
/usr/local/include/sdbus-c++/Message.h:93:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(bool&)
         Message& operator>>(bool& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:93:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘bool&’
/usr/local/include/sdbus-c++/Message.h:94:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(int16_t&)
         Message& operator>>(int16_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:94:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘int16_t& {aka short int&}’
/usr/local/include/sdbus-c++/Message.h:95:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(int32_t&)
         Message& operator>>(int32_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:95:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘int32_t& {aka int&}’
/usr/local/include/sdbus-c++/Message.h:96:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(int64_t&)
         Message& operator>>(int64_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:96:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘int64_t& {aka long int&}’
/usr/local/include/sdbus-c++/Message.h:97:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(uint8_t&)
         Message& operator>>(uint8_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:97:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘uint8_t& {aka unsigned char&}’
/usr/local/include/sdbus-c++/Message.h:98:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(uint16_t&)
         Message& operator>>(uint16_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:98:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘uint16_t& {aka short unsigned int&}’
/usr/local/include/sdbus-c++/Message.h:99:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(uint32_t&)
         Message& operator>>(uint32_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:99:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘uint32_t& {aka unsigned int&}’
/usr/local/include/sdbus-c++/Message.h:100:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(uint64_t&)
         Message& operator>>(uint64_t& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:100:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘uint64_t& {aka long unsigned int&}’
/usr/local/include/sdbus-c++/Message.h:101:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(double&)
         Message& operator>>(double& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:101:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘double&’
/usr/local/include/sdbus-c++/Message.h:102:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(char*&)
         Message& operator>>(char*& item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:102:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘char*&’
/usr/local/include/sdbus-c++/Message.h:103:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(std::__cxx11::string&)
         Message& operator>>(std::string &item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:103:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘std::__cxx11::string& {aka std::__cxx11::basic_string<char>&}’
/usr/local/include/sdbus-c++/Message.h:105:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(sdbus::ObjectPath&)
         Message& operator>>(ObjectPath &item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:105:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘sdbus::ObjectPath&’
/usr/local/include/sdbus-c++/Message.h:106:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(sdbus::Signature&)
         Message& operator>>(Signature &item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:106:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘sdbus::Signature&’
/usr/local/include/sdbus-c++/Message.h:107:18: note: candidate: sdbus::Message& sdbus::Message::operator>>(sdbus::UnixFd&)
         Message& operator>>(UnixFd &item);
                  ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:107:18: note:   no known conversion for argument 1 from ‘sdbus::MethodCall’ to ‘sdbus::UnixFd&’
/usr/local/include/sdbus-c++/Message.h:314:21: note: candidate: template<class _Element> sdbus::Message& sdbus::operator>>(sdbus::Message&, std::vector<_Element>&)
     inline Message& operator>>(Message& msg, std::vector<_Element>& items)
                     ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:314:21: note:   template argument deduction/substitution failed:
/usr/local/include/sdbus-c++/Message.h:370:13: note:   ‘sdbus::MethodCall’ is not derived from ‘std::vector<_Element>’
             (void)(msg >> ... >> args);
/usr/local/include/sdbus-c++/Message.h:336:21: note: candidate: template<class _Key, class _Value> sdbus::Message& sdbus::operator>>(sdbus::Message&, std::map<_Key, _Value>&)
     inline Message& operator>>(Message& msg, std::map<_Key, _Value>& items)
                     ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:336:21: note:   template argument deduction/substitution failed:
/usr/local/include/sdbus-c++/Message.h:370:13: note:   ‘sdbus::MethodCall’ is not derived from ‘std::map<_Key, _Value>’
             (void)(msg >> ... >> args);
/usr/local/include/sdbus-c++/Message.h:383:21: note: candidate: template<class ... _ValueTypes> sdbus::Message& sdbus::operator>>(sdbus::Message&, sdbus::Struct<_ValueTypes ...>&)
     inline Message& operator>>(Message& msg, Struct<_ValueTypes...>& item)
                     ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:383:21: note:   template argument deduction/substitution failed:
/usr/local/include/sdbus-c++/Message.h:370:13: note:   ‘sdbus::MethodCall’ is not derived from ‘sdbus::Struct<_ValueTypes ...>’
             (void)(msg >> ... >> args);
/usr/local/include/sdbus-c++/Message.h:400:21: note: candidate: template<class ... _ValueTypes> sdbus::Message& sdbus::operator>>(sdbus::Message&, std::tuple<_Tps ...>&)
     inline Message& operator>>(Message& msg, std::tuple<_ValueTypes...>& item)
                     ^~~~~~~~
/usr/local/include/sdbus-c++/Message.h:400:21: note:   template argument deduction/substitution failed:
/usr/local/include/sdbus-c++/Message.h:370:13: note:   ‘sdbus::MethodCall’ is not derived from ‘std::tuple<_Tps ...>’
             (void)(msg >> ... >> args);

```

**Comment by sangelovic** [comment]
1. Yes, since D-Bus is a messaging system with queues, sdbus-c++ server will queue up incoming messages (calls, signals, replies...) and these messages are handled sequentially, one by one, by the event loop. So when clients issue calls B and C in parallel while your D-Bus server is busy handling call A, then B and C will be queued up at the server side to be handled right after A has been handled. This is taken care by the `enterEventLoop();`, or `enterEventLoopAsync();`, respectively. The only difference between them is that the former one blocks the thread that called it (and that thread thus becomes the thread handling incoming D-Bus messages, one by one, and will not return from the `enterEventLoop();` until you stop the event loop explicitly), while the latter one spawns another thread to handle incoming message, and your current thread will continue after the `enterEventLoopAsync();` call (i.e. your current thread is not stuck inside the call).

2. No, D-Bus is not shared-memory IPC, but a message-based IPC. What you pass to another communication partner is a message. In that message must be serialized everything you want to pass to him. The message (including all data) is then trasmitted through sockets. If you need to convey large chunks of data efficiently, I'd first measure. If D-Bus does not fulfill your requirements, you should look around for other IPCs more fit for your use case.

**Comment by vk-gst** [comment]
Thanks for the update. 

The max. size that I would pass to-n-fro between client-server is in the range of 3500-4000 bytes. The calls will be asynchronous but they wont occur every second. 

As per DBus specification, the bandwidth is around 64MB. For my use case, would Dbus IPC it still suffice, or I should look for some other IPCs?

**Comment by sangelovic** [comment]
I can't tell, it may depend on your system. But generally I guess it should. You can try and see.
